pkg/migration-tool/reader/reader.go

```diff
@@ -0,0 +1,174 @@
+}
+
+// New creates a new RemoteRead. It creates a ReadClient that is imported from Prometheus remote storage.
+// RemoteRead takes help of plan to understand how to create fetchers.
+func New(readStorageUrl string, p *plan.Plan, sigRead, sigWrite chan struct{}) (*RemoteRead, error) {
+	parsedUrl, err := url.Parse(readStorageUrl)
+	if err != nil {
+		return nil, fmt.Errorf("url-parse: %w", err)
+	}
+	rc, err := remote.NewReadClient(fmt.Sprintf("reader-%d", 1), &remote.ClientConfig{
+		URL:     &config.URL{URL: parsedUrl},
+		Timeout: model.Duration(DefaultReadTimeout),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating remote-client: %w", err)
+	}
+	read := &RemoteRead{
+		url:           readStorageUrl,
+		plan:          p,
+		client:        rc,
+		sigBlockRead:  sigRead,
+		sigBlockWrite: sigWrite,
+	}
+	read.fetchersRunning.Store(0)
+	return read, nil
+}
+
+// Run runs the remote read and starts fetching the samples from the read storage.
+func (rr *RemoteRead) Run(wg *sync.WaitGroup, readerUp *atomic.Bool) error {
+	defer func() {
+		close(rr.sigBlockRead)
+		close(rr.sigBlockWrite)
+		readerUp.Store(false)
+		wg.Done()
+	}()
+	readerUp.Store(true)
+	log.Info("msg", "reader is up")
+	timeRangeMinutesDelta := time.Minute.Milliseconds()
+	for i := rr.plan.Mint; i <= rr.plan.Maxt; {
+		var (
+			mint     = i
+			maxt     = mint + timeRangeMinutesDelta
+			fetch    = rr.createFetch(rr.url, mint, maxt)
+			blockRef = rr.plan.CreateBlock(mint, maxt)
+		)
+		blockRef.SetDescription(fmt.Sprintf("fetching time-range: %d mins...", timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		result, err := fetch.start(context.Background(), []*labels.Matcher{labels.MustNewMatcher(labels.MatchRegexp, "", ".*")})
+		if err != nil {
+			return fmt.Errorf("remote-read run: %w", err)
+		}
+		if len(result.Timeseries) == 0 {
+			rr.plan.DecrementBlockCount()
+			i += timeRangeMinutesDelta + 1
+			if timeRangeMinutesDelta < MaxTimeRangeDeltaLimit.Milliseconds() {
+				timeRangeMinutesDelta *= 2
+			}
+			continue
+		}
+		qResultSize := result.Size() // Assuming each character corresponds to one byte.
+		rr.plan.UpdateLastPushedMaxt(i)
+		blockRef.SetDescription(fmt.Sprintf("received %d MB with delta %d mins...", qResultSize/megaByte, timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		blockRef.Timeseries = result.Timeseries
+		rr.sigBlockRead <- struct{}{}
+		<-rr.sigBlockWrite
+		if err := rr.plan.Clear(); err != nil {
+			return fmt.Errorf("remote-read run: %w", err)
+		}
+		switch {
+		case qResultSize < ResponseDataSizeHalfLimit && timeRangeMinutesDelta >= MaxTimeRangeDeltaLimit.Milliseconds():
+			i += timeRangeMinutesDelta + 1
+		case qResultSize < ResponseDataSizeHalfLimit:
+			i += timeRangeMinutesDelta + 1
+			timeRangeMinutesDelta *= 2
+		default:
+			i += timeRangeMinutesDelta + 1
+			timeRangeMinutesDelta /= 2
+		}
+	}
+	log.Info("msg", "reader is down")
+	return nil
+}
+
+type fetch struct {
```

Comment:
the fetcher in the current implementation seems like it is not needed, it is only used once to call the start method and discarded, can be a function

Reply:
For now, yes that is best. But I think the reason why this was done is that fetch was planned to be able to divide 1 big request into multiple small requests and hence this was created in the way it is. Now, I don't see that coming up, so better to just keep this as a function. Thank you.